[homematic] Migration of homematic binding

mdicke2s opened the pull request:
As previously discussed between @kaikreuzer and @gerrieg, it is intended to move the homematic binding from the Openhab project to Eclispe Smarthome. This commit migrates the sources of the binding to ESH repository.

**Details of the original binding:**
- Source: https://github.com/openhab/openhab2-addons/tree/master/addons/binding/org.openhab.binding.homematic/
- Initial contributed by: @gerrieg 
- License: EPL v1.0

**The following steps were applied:**
- Copied classes and resources from Openhab
- Updated settings and documentation which used the term "openhab"
- Renamed project and namespaces
- Added project to maven
- updated copyright headers using `mvn license:format`
- Locally built JAR of migrated binding was tested successfully with a few homematic devices on an ESH installation.

**Next steps:**
- Move open issues to ESH issue tracking (as long as they are still valid)
- Remove binding's code from Openhab codebase

**Contributions on the binding**

| Github user      | Full Name             | Email                              | Contribution |
| ---------------- | --------------------- | ---------------------------------- | ------------ |
| _Initial by:_    |                       |                                    |              |
| @gerrieg         | Gerhard Riegler       | gerhard.riegler@gmail.com          | has ECA      |
| _Also by:_       |                       |                                    |              |
| @kaikreuzer      | Kai Kreuzer           | kai@openhab.org                    | has ECA      |
| @martinvw        | Martin van Wingerden  | martinvw@mtin.nl                   | has ECA      |
| @FStolte         | Florian Stolte        | fstolte@itemis.de                  | has ECA      |
| @wborn           | Wouter Born           | eclipse@maindrain.net              | has ECA      |
| @svilenvul       | Svilen Valkanov       | svilen.valkanov@musala.com         | has ECA      |
| @mdicke2s        | Michael Reitler       | michael.reitler@telekom.de         | has ECA      |
| @MHerbst         | Martin Herbst         | mail@mherbst.de                    | has ECA      |
| @triller-telekom | Stefan Triller        | stefan.triller@telekom.de          | has ECA      |
| @VelinYordanov   | VelinYordanov         | velin.iordanov@gmail.com           | has ECA      |
| @bgilmer77       | Brad Gilmer           | brad@gilmer.tv                     | has ECA      |
| @cweitkamp       | Christoph Weitkamp    | github@christophweitkamp.de        | has ECA      |
| @Confectrician   | Jerome Luckenbach     | github@luckenba.ch                 | has ECA      |
| @marcelrv        | Marcel Verpaalen      | marcel@verpaalen.com               | has ECA      |
| @maggu2810       | Markus Rathgeb        | maggu2810@gmail.com                | has ECA      |
| @3yc             | Adam Majerczyk        | kontakt.eclipse@3yc.de             | has ECA      |
| @maniac103       | Danny Baumann         | dannybaumann@web.de                | has ECA |
| @alexreinert     | Alexander Reinert     | alex@areinert.de                   | has ECA |
| @pauxus          | Stephan Pauxberger    | stephan@blackbuild.com             | has ECA |
| _Not mentioned in commit (but only contributed trivial change which is not covered by copyright):_       |                       |                                    |              |
| @falkena         | Alexander Falkenstern | falkena@users.noreply.github.com   | only trivial change, see openhab issue openhab/openhab2-addons#2221 |
| @kirilAtan       | Kiril Atanasov        | mail@kisat.eu                      | only trivial change, see openhab issue openhab/openhab2-addons#1263 |
| @bullshit        | Oskar                 | bullshit@users.noreply.github.com  | only trivial change, see openhab issue openhab/openhab2-addons#1771 |
| @tholtgre        | tholtgre              | thomas@holtgrefe.net               | only trivial change, see openhab issue openhab/openhab2-addons#1291 |

Signed-off-by: Michael Reitler <michael.reitler@telekom.de>

mdicke2s commented:
@gerrieg Would you mind to take a look on the open issues https://github.com/openhab/openhab2-addons/labels/homematic ? Probably some of them are either deprecated or can be closed as duplicates. Since you have the best knowledge of the binding, I would appreciate your help to avoid creating irrelevant issues in ESH.

kaikreuzer commented:
> Initial contributed by: @gerrieg

Please note that you have to mention him (and all other contributors to the binding's code) with an "Also-By" in the commit message - see https://github.com/eclipse/smarthome/blob/master/CONTRIBUTING.md#technical-considerations.

kaikreuzer commented:
Please also make sure to fully align it with the other ESH bundles wrt EPLv2 (e.g. the license in the manifest, the LICENSE file, etc.)

mdicke2s commented:
@kaikreuzer I updated the PR

1. Thanks for reminding me of the "Also-By" clause - actually I read over it but I did't think about it when I committed. Therefore I extracted the committers of the binding from git history.
2. After you reminded me, I doublechecked which files may get updated with the EPLv2. It got updated in the Manifest (was still 1.0) and I added a NOTICE.txt which contains license infos. Other than that I do not see special files in other bindings (e.g. didn't find any LICENSE file - what did you mean by that?)

pauxus commented:
Technical idea: 
would it not make more sense to copy the binding over with (at least some of) the history?

I.e.:
- clone the openhabe repository, 
- do a subdirectory filter, 
- move the files into the correct folder of the ESH structure
- push the branch as orphan branch into ESH (or a fork of ESH)
- merge the ESH master into the new orphan branch
- do the additional  necessary fixes
- Create the pull request and eventually merge the branch into ESH proper.

That would a) preserve the history (which the binding has a lot of) and b) should also avoid the need for the long also-by list, since individual commits would still be linked to their respective authors.

kaikreuzer commented:
@pauxus No, this won't work - we have to get this through the Eclipse IP process and for this it needs a single clean commit, which will be triaged.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.homematic/NOTICE.txt:
```diff
@@ -0,0 +1,19 @@
+This content is produced and maintained by the Eclipse SmartHome project.
```
The filename should be "NOTICE", not "NOTICE.txt", see e.g. https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.astro/NOTICE

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.homematic/README.md:
```diff
@@ -0,0 +1,498 @@
+## Supported Things
+
+All devices connected to a Homematic gateway.
+All required ESH metadata are generated during device discovery.
```
Avoid "ESH" as it is no official abbreviation and it should not even turn up in solutions documentation.
Best simply remove ESH here.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.homematic/README.md:
```diff
@@ -0,0 +1,498 @@
+Hint for the binding to identify the gateway type (auto|ccu|noccu) (default = auto).
+
+-   **callbackHost**  
+Callback network address of the ESH runtime, default is auto-discovery
```
Avoid "ESH" as it is no official abbreviation and it should not even turn up in solutions documentation.
Best simply remove ESH here.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.homematic/src/main/java/org/eclipse/smarthome/binding/homematic/handler/HomematicBridgeHandler.java:
```diff
@@ -0,0 +1,367 @@
+                gateway.startWatchdogs();
+            } catch (IOException ex) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, ex.getMessage());
+                logger.error("Homematic bridge was set to OFFLINE-COMMUNICATION_ERROR due to the following exception: {}",
```
this should be DEBUG

kaikreuzer commented:
Wrt the author list, this is pretty long already 😟 .

Note that we need valid e-mail addresses xxx@users.noreply.github.com won't be accepted.
Ideally, all contributors should have signed an ECA - you can check, who already has an ECA on https://accounts.eclipse.org/users/kkreuzer/eca; for all others, we should ask them if they could sign it or alternatively check whether their contribution was only a "small patch" (e.g. single line fixes like typos) or could be easily rewritten by us (you :-)).

Confectrician commented:
>Note that we need valid e-mail addresses[...]

I will make that easier for my personal case

@Confectrician Jerome Luckenbach github@luckenba.ch

ECA signed already some time ago.

martinvw commented:
And mine should be martinvw@mtin.nl there should be an ECA for that one

vbier commented:
I can be removed from the list as  my change has been reverted.

3yc commented:
My changes are quite small.
Anyhow, ECA is signed, my email: kontakt.eclipse@3yc.de

> Am 11.06.2018 um 14:59 schrieb Kai Kreuzer <notifications@github.com>:
> 
> Wrt the author list, this is pretty long already 😟 .
> 
> Note that we need valid e-mail addresses xxx@users.noreply.github.com <mailto:xxx@users.noreply.github.com> won't be accepted.
> Ideally, all contributors should have signed an ECA - you can check, who already has an ECA on https://accounts.eclipse.org/users/kkreuzer/eca <https://accounts.eclipse.org/users/kkreuzer/eca>; for all others, we should ask them if they could sign it or alternatively check whether their contribution was only a "small patch" (e.g. single line fixes like typos) or could be easily rewritten by us (you :-)).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/eclipse/smarthome/pull/5695#issuecomment-396234700>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEWPNZVE37g943aJ0A2HT3lkvyITsR6Pks5t7mnNgaJpZM4UeSFe>.
>

mdicke2s commented:
@maniac103 Could you please agree to the Eclipse Contributor Agreement in your eclipse.org account and let me know the email address that you used there? I would appreciate that a lot.

mdicke2s commented:
@alexreinert, @falkena, @kirilAtan, @bullshit, @pauxus, @tholtgre: For your mail addresses I couldn't find an Eclipse Contributor Agreement on eclipse.org. Would you mind to sign and ECA and let me know which email address you used for that?

**Remark:** The contributions that you made are rather small (additions to README, nullpointer checks, etc.). On the first look they appear to be trivial (e.g. there isn't many ways of doing a nullpointer check, so it could be easily rewritten by somebody else). Therefore I may rewrite your contribution and not mention you in the commit if you do not sign an ECA.

alexreinert commented:
Just signed the ECA.

maniac103 commented:
Just signed the ECA as well. Mail address is the one above (dannybaumann@web.de).

pauxus commented:
Signed (stephan@blackbuild.com)

falkena commented:
I can be removed from the list, since only homegear NPE fixed.

mdicke2s commented:
@kaikreuzer Thanks for your feedback. As you can see I managed to clarify the license topic. Furthermore I addressed your findings in the second commit. Could you please check again?

Currently the PR is spitted up in 3 commits. The firs one is mainly copy&rename and the other two are things that I did on top. I left it like that because I guess it is optimal for reviewing (this first commit is mainly the one which comes from multiple contributors on openhab project). If it is preferred to squash them, please let me know.

kaikreuzer commented:
Thanks @mdicke2s Looks all good!
And yes, for creating the CQ, we will need the commits squashed into one, so please do so. I'll then create the CQ and from that moment on you must not touch the branch anymore until it gets merged.

mdicke2s commented:
Good news @kaikreuzer 
Commits were squashed, from now on I am not going touch the branch any more

kaikreuzer commented:
Perfect, thank you!
I have created CQ 16751 and we will now wait for approval.

kaikreuzer commented:
Good news, the CQ has been approved for check-in!

kaikreuzer commented:
@gerrieg I'd plan to do a new ESH stable release next Tuesday (June 26) and will then remove the binding from openhab2-addons and adapt the openHAB distro accordingly, so that it picks up the ESH version. There should hopefully be no noticable impact on existing users through that change.

Feel free to already do PRs against the binding here, if anything was hold back in the past weeks.

wborn commented:
What would be the best way to migrate the [22 open Homematic issue](https://github.com/openhab/openhab2-addons/issues?q=is%3Aopen+is%3Aissue+label%3Ahomematic)?

mdicke2s commented:
@wborn I have that on my list. The idea is to use https://github.com/google/github-issue-mover which seems reliable to me. That way we can keep discussions from the existing issues without much manual effort.

mdicke2s commented:
Meanwhile I applied the Github-issue-mover to the following issues
https://github.com/openhab/openhab2-addons/issues/3637
https://github.com/openhab/openhab2-addons/issues/3567
https://github.com/openhab/openhab2-addons/issues/3559
https://github.com/openhab/openhab2-addons/issues/3507
https://github.com/openhab/openhab2-addons/issues/3309
https://github.com/openhab/openhab2-addons/issues/3307
https://github.com/openhab/openhab2-addons/issues/3177
https://github.com/openhab/openhab2-addons/issues/3166
https://github.com/openhab/openhab2-addons/issues/3153
https://github.com/openhab/openhab2-addons/issues/3068
https://github.com/openhab/openhab2-addons/issues/3026
https://github.com/openhab/openhab2-addons/issues/2849
https://github.com/openhab/openhab2-addons/issues/2817
https://github.com/openhab/openhab2-addons/issues/2803
https://github.com/openhab/openhab2-addons/issues/2568
https://github.com/openhab/openhab2-addons/issues/2543
https://github.com/openhab/openhab2-addons/issues/2436
https://github.com/openhab/openhab2-addons/issues/2247
https://github.com/openhab/openhab2-addons/issues/1978
https://github.com/openhab/openhab2-addons/issues/1908
https://github.com/openhab/openhab2-addons/issues/1668
https://github.com/openhab/openhab2-addons/issues/1580

@kaikreuzer: Since my github user does not have the permission to close them on the openhab project, could you please do so? They do already have a comment that they were moved, so there is not need for further documentation.

martinvw commented:
@mdicke2s / @kaikreuzer should be all closed now.

mdicke2s commented:
Perfect :-) Thx @martinvw

kaikreuzer commented:
> @gerrieg I'd plan to do a new ESH stable release next Tuesday (June 26) and will then remove the binding from openhab2-addons and adapt the openHAB distro accordingly

FTR: https://github.com/openhab/openhab2-addons/pull/3701 and https://github.com/openhab/openhab-distro/pull/730.

kaikreuzer commented:
FTR: I have reverted these changes for now as https://github.com/eclipse/smarthome/pull/5820 is still open and the ESH build currently has major issues (probably due to server load through Photon).
